sycl/source/detail/scheduler/commands.cpp

```diff
@@ -161,6 +226,190 @@ Command::Command(CommandType Type, QueueImplPtr Queue)
   MEvent->setContextImpl(detail::getSyclObjImpl(MQueue->get_context()));
+
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  // Obtain the stream ID so all commands can emit
+  // traces to that stream
+  if (xptiTraceEnabled()) {
+    MStreamID = xptiRegisterStream(SYCL_STREAM_NAME);
+  }
+#endif
+}
+
+void Command::emitInstrumentationDataProxy() {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  emitInstrumentationData();
+#endif
+}
+
+void Command::emitEdgeEventForCommandDependence(Command *Cmd, void *ObjAddr,
+                                                const string_class &Prefix,
+                                                bool IsCommand) {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  // If all the information we need for creating an edge event is available,
+  // then go ahead with creating it; if not, bail early!
+  if (xptiTraceEnabled() && MTraceEvent && Cmd && Cmd->MTraceEvent) {
+    xpti::utils::StringHelper SH;
+    std::string AddressStr = SH.addressAsString(ObjAddr);
+    std::string TypeString = SH.nameWithAddressString(Prefix, AddressStr);
+    // Create an edge with the dependent buffer address being one of the
+    // properties of the edge
+    xpti::payload_t Payload(TypeString.c_str(), MAddress);
+    auto EdgeEvent = xptiMakeEvent(TypeString.c_str(), &Payload,
+                                   (uint16_t)xpti_te::graph, xpti_at::active);
+    // Bail early if either the source or the target node
+    // for the given dependency is undefined or NULL
+    if (EdgeEvent && MTraceEvent) {
+      xpti_td *SrcEvent = (xpti_td *)Cmd->MTraceEvent;
+      xpti_td *TgtEvent = (xpti_td *)MTraceEvent;
+      EdgeEvent->source_id = SrcEvent->unique_id;
+      EdgeEvent->target_id = TgtEvent->unique_id;
+      if (IsCommand) {
+        xptiAddMetadata(EdgeEvent, "access_mode", TypeString.c_str());
+        xptiAddMetadata(EdgeEvent, "memory_object", AddressStr.c_str());
+      } else {
+        xptiAddMetadata(EdgeEvent, "event", TypeString.c_str());
+      }
+      xptiNotifySubscribers(MStreamID, (uint16_t)xpti_tp::edge_create,
+                            cl::sycl::detail::g_sycl_graph_event, EdgeEvent,
+                            nullptr);
+    }
+    // General comment - should we print error message for odd/bad
+    // events or ignore them for now?
+  }
+#endif
+}
+
+void Command::emitEdgeEventForEventDependence(Command *Cmd,
+                                              RT::PiEvent &PiEventAddr) {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  // If we have failed to create an event to represent the Command, then we
+  // cannot emit an edge event. Bail early!
+  if (!MTraceEvent)
+    return;
+
+  if (xptiTraceEnabled()) {
+    if (Cmd && Cmd->MTraceEvent) {
+      // If the event is associated with a command, we use this command's trace
+      // event as the source of edge, hence modeling the control flow
+      emitEdgeEventForCommandDependence(Cmd, (void *)PiEventAddr, "Event",
+                                        false);
+
```

Comment:
Please restructure the code so:

```
if(smt) {
  do_smt();
  return;
}
if(smt) {
  do_smt();
  return;
}
do_smt();

```

See: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Please, apply to whole patch.

Reply:
Will do this where ever it makes sense.

Reply:
Done!